sql: support version numbers on descriptor validation

fqazi opened the pull request:
Previously, the descriptor validation code did not
take a version number, so it was not possible to version
gate new validation logic. This was inadequate  because
when new fields are introduced we don't want their validation
to kick in for certain cases like the debug doctor, such as any
new fields with non-zero defaults. To address this, this patch add supports
for version numbers inside the validation, and updates unit tests
to pass this in as well. It also adds a new option on debug doctor
to run a version of validation.

Release note (cli change): Add new optional version argument
to the doctor examine command. This can be used to enable /
disable validation when examining older zip directories.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76166)
<!-- Reviewable:end -->

ajwerner reviewed (commented):
> Reviewed 54 of 146 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76166)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @fqazi, and @postamar)

___
*[pkg/sql/catalog/descs/descriptor.go, line 297 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76166#-MvK5JyT9UmLNUD2pOtQ:-MvK5JyU76-1kmk1k_Rx:bvneyre) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc9774033587dd185af0e17f24220fcd7f890e43/pkg/sql/catalog/descs/descriptor.go#L297)):*
> ```Go
> 	descs, err = tc.withReadFromStore(mutable, func() ([]catalog.MutableDescriptor, error) {
> 		uncommittedDB, _ := tc.uncommitted.getByID(parentID).(catalog.DatabaseDescriptor)
> 		version := tc.settings.Version.ActiveVersion(ctx)
> ```

shouldn't this use the version we already initialized and stored in the collection?
___
*[pkg/sql/catalog/desctestutils/descriptor_test_utils.go, line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76166#-MvKAVT986_F1Mn2alh2:-MvKAVT986_F1Mn2alh3:buz29et) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc9774033587dd185af0e17f24220fcd7f890e43/pkg/sql/catalog/desctestutils/descriptor_test_utils.go#L35)):*
> ```Go
> // TestingGetDatabaseDescriptor retrieves a database descriptor directly from
> // the kv layer.
> func TestingGetDatabaseDescriptor(
> ```

Having reviewed a lot of this change, I wonder to what extent this plumbing here is making things a lot more painful. Should we make a variant which takes a version and just hard-code it to the latest version in the existing variant and plumb that through to the new variant? It feels like it would reduce the number of changes dramatically. 


<!-- Sent from Reviewable.io -->

fqazi committed:
sql: support version numbers on descriptor validation #76166

Previously, the descriptor validation code did not
take a version number, so it was not possible to version
gate new validation logic. This was inadequate because
when new fields are introduced we don't want their validation
to kick in for certain cases like the debug doctor, such as any
new fields with non-zero defaults. To address this, this patch add supports
for version numbers inside the validation, and updates unit tests
to pass this in as well. It also adds a new option on debug doctor
to run a version of validation.

Release note (cli change): Add new optional version argument
to the doctor examine command. This can be used to enable /
disable validation when examining older zip directories.

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76166)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @postamar)

___
*[pkg/sql/catalog/desctestutils/descriptor_test_utils.go, line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76166#-MvKAVT986_F1Mn2alh2:-MvKpVUo1RQAnKQjVPcm:b-j8u1xh) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc9774033587dd185af0e17f24220fcd7f890e43/pkg/sql/catalog/desctestutils/descriptor_test_utils.go#L35)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Having reviewed a lot of this change, I wonder to what extent this plumbing here is making things a lot more painful. Should we make a variant which takes a version and just hard-code it to the latest version in the existing variant and plumb that through to the new variant? It feels like it would reduce the number of changes dramatically. 
</blockquote></details>

Done.

Made this the default behaviour which reduces the number of code changes by a lot.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 22 of 146 files at r1, 1 of 1 files at r2, 80 of 108 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76166#-:-MvKvMHp3BirHCTQ_jqq:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @fqazi, and @postamar)

___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 897 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76166#-MvKu-Ii5Fz1TOAfsaQb:-MvKu-Ii5Fz1TOAfsaQc:bvwd07z) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d3c8a9e040c96d33ed4be394ff0943b232dd1b7/pkg/ccl/changefeedccl/changefeed_stmt.go#L897)):*
> ```Go
> 	ctx context.Context, execCfg *sql.ExecutorConfig, txn *kv.Txn, desc catalog.TableDescriptor,
> ) (string, error) {
> 	col := execCfg.CollectionFactory.MakeCollection(ctx, nil)
> ```

nit: add back comment
___
*[pkg/ccl/changefeedccl/rowfetcher_cache.go, line 74 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76166#-MvKu0mY9dqgDa50gqpS:-MvKu0mY9dqgDa50gqpT:bvwd07z) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef0fede944f7a9f132bc11c29640b6041abefaa/pkg/ccl/changefeedccl/rowfetcher_cache.go#L74)):*
> ```Go
> 		codec:      codec,
> 		leaseMgr:   leaseMgr,
> 		collection: cf.NewCollection(nil /* TemporarySchemaProvider */),
> ```

nit: add back comment
___
*[pkg/cli/doctor.go, line 96 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76166#-MvKuGm85y0FGQ8tFo5b:-MvKuGm85y0FGQ8tFo5c:b2fbbep) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d3c8a9e040c96d33ed4be394ff0943b232dd1b7/pkg/cli/doctor.go#L96)):*
> ```Go
> that may not exist on downlevel versions.
> `,
> 		Args: cobra.MinimumNArgs(1),
> ```

nit: `cobra.RangeArgs(1, 2)`
___
*[pkg/sql/create_view.go, line 209 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76166#-MvKuZoh-3Zlr_C_ydjM:-MvKuZoh-3Zlr_C_ydjN:b-gbwqh7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d3c8a9e040c96d33ed4be394ff0943b232dd1b7/pkg/sql/create_view.go#L209)):*
> ```Go
> 		// currently relied on in import and restore code and tests.
> 		var creationTime hlc.Timestamp
> 		desc, err := makeViewTableDesc(params.ctx, viewName, n.viewQuery, n.dbDesc.GetID(), schema.GetID(), id, n.columns, creationTime, privs, &params.p.semaCtx, params.p.EvalContext(), params.p.EvalContext().Settings, n.persistence, n.dbDesc.IsMultiRegion(), params.p)
> ```

nit: consider wrapping this
___
*[pkg/sql/temporary_schema.go, line 524 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76166#-MvKucdf8FhMsmIOtzwR:-MvKucdf8FhMsmIOtzwS:bvwd07z) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d3c8a9e040c96d33ed4be394ff0943b232dd1b7/pkg/sql/temporary_schema.go#L524)):*
> ```Go
> 	// Build a set of all databases with temporary objects.
> 	var allDbDescs []catalog.DatabaseDescriptor
> 	descsCol := c.collectionFactory.NewCollection(ctx, nil)
> ```

nit: add back comment
___
*[pkg/sql/virtual_schema.go, line 239 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76166#-MvKuhIz7mLzN37Iz_0K:-MvKuhIz7mLzN37Iz_0L:by6fkfu) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d3c8a9e040c96d33ed4be394ff0943b232dd1b7/pkg/sql/virtual_schema.go#L239)):*
> ```Go
> ) (descpb.TableDescriptor, error) {
> 	stmt, err := parser.ParseOne(v.schema)
> ```

nit: newline
___
*[pkg/sql/virtual_schema.go, line 254 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76166#-MvKug--A4A6cgXIHFL3:-MvKug--A4A6cgXIHFL4:b-5kdkp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ef0fede944f7a9f132bc11c29640b6041abefaa/pkg/sql/virtual_schema.go#L254)):*
> ```Go
> 		create.Name.Table(),
> 		tree.AsStringWithFlags(create.AsSource, tree.FmtParsable),
> 		0, /* parentID */
> ```

nit: add back comments


<!-- Sent from Reviewable.io -->

fqazi commented:
bors r+
